ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/MLSdkAsyncHttpResponseHandler.java

```diff
@@ -0,0 +1,194 @@
+        Map<Integer, ModelTensors> tensorOutputs
+    ) {
+        if (Strings.isBlank(body)) {
+            log.error("Remote model response body is empty!");
+            if (executionContext.getExceptionHolder().get() == null)
+                executionContext
+                    .getExceptionHolder()
+                    .compareAndSet(null, new OpenSearchStatusException("No response from model", RestStatus.BAD_REQUEST));
+        } else {
+            if (statusCode < HttpStatus.SC_OK || statusCode > HttpStatus.SC_MULTIPLE_CHOICES) {
+                log.error("Remote server returned error code: {}", statusCode);
+                if (executionContext.getExceptionHolder().get() == null)
+                    executionContext
+                        .getExceptionHolder()
+                        .compareAndSet(null, new OpenSearchStatusException(REMOTE_SERVICE_ERROR + body, RestStatus.fromCode(statusCode)));
+            } else {
+                try {
+                    ModelTensors tensors = processOutput(body, connector, scriptService, parameters, mlGuard);
+                    tensors.setStatusCode(statusCode);
+                    tensorOutputs.put(executionContext.getSequence(), tensors);
+                } catch (Exception e) {
+                    log.error("Failed to process response body: {}", body, e);
+                    if (executionContext.getExceptionHolder().get() == null)
+                        executionContext
+                            .getExceptionHolder()
+                            .compareAndSet(null, new MLException("Fail to execute predict in aws connector", e));
+                }
+            }
+        }
+    }
+
+    // Only all requests successful case will be processed here.
+    private void reOrderTensorResponses(Map<Integer, ModelTensors> tensorOutputs) {
+        ModelTensors[] modelTensors = new ModelTensors[tensorOutputs.size()];
+        log.debug("Reordered tensor outputs size is {}", tensorOutputs.size());
+        for (Map.Entry<Integer, ModelTensors> entry : tensorOutputs.entrySet()) {
+            modelTensors[entry.getKey()] = entry.getValue();
+        }
+        actionListener.onResponse(Arrays.asList(modelTensors));
+    }
+
+    protected class MLResponseSubscriber implements Subscriber<ByteBuffer> {
+        private Subscription subscription;
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            this.subscription = s;
+            s.request(Long.MAX_VALUE);
+        }
+
+        @Override
+        public void onNext(ByteBuffer byteBuffer) {
+            responseBody.append(StandardCharsets.UTF_8.decode(byteBuffer));
+            subscription.request(Long.MAX_VALUE);
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            log
+                .error(
+                    "Error on receiving response body from remote: {}",
+                    t instanceof NullPointerException ? "NullPointerException" : t.getMessage(),
```

Comment:
the code assumes `t.getMessage()` will never be null. This will hold true in most cases, however, there is a chance that in certain exceptions/error conditions `t.getMessage(`) could be null 

```
t instanceof NullPointerException ? "NullPointerException" : (t.getMessage() != null ? t.getMessage() : "Unknown Error")
```

Reply:
Exception can be NullPointerException, and in that case if jdk version is under 9, it can cause NEP here and user won't able to get response.